### PR TITLE
Fix to 63130

### PR DESCRIPTION
**Pull request by sanjeevnec**
## Description
Class name: HTTPArgument.java
Method name: HTTPArgument (constructor)
Class name: HttpHC4Impl.java
Method name: createUrlEncodedFormEntity

Replaced java.net.URLDecoder class from org.apache.commons.codec.net.URLCodec class

## Motivation and Context
JMeter uses java.net.URLDecoder class for URL encoding.
Its drawback : This library can not encode Shift_JIS correctly because of different specific interpretation between Java and IE.
Because of browser specification, char code is zipped.
Browser encode Japanese text as following.
初期値大阪市 ⇒ %8F%89%8A%FA%92l%91%E5%8D%E3%8Es
初 = %8F%89
期 = %8A%FA
値 = %92l ← % is lack
大 = %91%E5
阪 = %8D%E3
市 = %8Es ← % is lack

JMeter should use org.apache.commons.codec.net.URLCodec class for URL encoding.

https://bz.apache.org/bugzilla/show_bug.cgi?id=63130

## How Has This Been Tested?
I have created one sample application, which contain some pages having shift_jis encoding, and deployed the same in tomcat server.
Created the test plan in Jmeter tool and configure it so that it can receive all the incoming request from the application.
Start the Http(s) Test Script Recorder
Run the application, and post the request from the application.
Validate the result in response page.
Run the "ant test" command to be ensured that no existing tests are breaking because of these changes.
Run the "ant package-and-check" to check the coding guidelines.

Other Details:

OS: windows 8
Java: 8
Jmeter version: 5.0
Hardware: PC

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
[x ] My code follows the code style of this project.
 I have updated the documentation accordingly.

### CODE REVIEW

File: test/resources/org/apache/jmeter/protocol/http/util/shift_jis.properties [hunk 1]
```diff
@@ -0,0 +1,2 @@
+value=����l���s
```
Comment (vlsi) [inline test/resources/org/apache/jmeter/protocol/http/util/shift_jis.properties]: Please avoid adding "non-utf-8" files to the repository.
It complicates support.

**Review by vlsi — CHANGES_REQUESTED** [review]

File: test/resources/org/apache/jmeter/protocol/http/util/shift_jis.properties (hunk 1 above)
Comment (sanjeevnec) [reply test/resources/org/apache/jmeter/protocol/http/util/shift_jis.properties]: This was added in order to test the functionality of encoding via "shiftjis" which is added via this PR.

Comment (vlsi) [reply test/resources/org/apache/jmeter/protocol/http/util/shift_jis.properties]: The are lots of way to test encoding. You don't have to create files for that.
You could just use `new byte[]{1,2,3,4}`

File: test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java [hunk 2]
```diff
@@ -19,79 +19,199 @@
...
-            arg.setAlwaysEncoded(false);
-            assertEquals("name.?", arg.getEncodedName());
-            assertEquals("value_ here", arg.getEncodedValue());
-            
-            arg = new HTTPArgument("", "\00\01\07", "", false);
-            arg.setAlwaysEncoded(false);
-            assertEquals("", arg.getEncodedName());
-            assertEquals("\00\01\07", arg.getEncodedValue());
-        }
+	@Rule
```
Comment (ham1) [inline test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java]: Please use spaces not tabs for indentation.

Comment (sanjeevnec) [reply test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java]: Incorporated the review comments for changing the tabs to spaces

Comment (sanjeevnec) [reply test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java]: Hi,

We had raised PR#440 which was closed due to migration of SVN>GIT.
So now we have raised a new PR#467 for  Bug#63130. 
Also all the comments which were raised for PR#440 have been incorporated in this new PR#467.
Kindly confirm those comments in this new PR#467.

Also, in PR#440, there was 1 comment (Mentioned below) related to encoding requirement for parameter name. 
"In this code you use a different encoding than UTF-8. Are you sure it's allowed for parameter names ? If yes, then the PR should impact getEncodedName() which uses UTF-8 by default".

As per our understanding & discussion with one of the Japanese developer, name is having a static value in the page and it has been used by the developers in the English only.  

So just want to have some input from other community members also.

Thank you!

File: test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java [hunk 3]
```diff
@@ -88,10 +96,120 @@ public void testEncoding() throws Exception {
...
+            assertEquals("arg.getEncodedName() should be equal to name.?", "name.?", arg.getEncodedName());
+            assertEquals("arg.getValue() should be equal to value_here", "value_ here", arg.getEncodedValue());
+    
+            // When the name and value parameter does not have valid encoded value (Encoded
+            // and Decoded values are same)
+            // In real encoded value for "name.?" is "name.%3F" and the encoded value for
+            // "value_here" is "value_+here"
+            arg = new HTTPArgument("name.?", "value_ here", true, encoding);
+    
+            assertEquals("arg.getName() should be equal to name.?", "name.?", arg.getName());
```
Comment (vlsi) [inline test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java]: It is great you add assert messages, however it makes little sense to duplicate "expected" value in the message.

In case of failure, the message would include "expected" value anyway, so phrases like `should be equal to name.?` add nothing.

You should probably use something like `assertEquals("HTTPArgument(\"name.?\", \"value_ here\",...).getName()", "name.?", arg.getName());`

Then it would produce clear message and tell what was the input, the action, the expected output and the actual output.

File: test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java [hunk 4]
```diff
@@ -88,10 +96,120 @@ public void testEncoding() throws Exception {
...
+    
+            // Try a real encoded argument
+            arg = new HTTPArgument("name.%3F", "value_+here", true, encoding);
+            assertEquals("arg.getName() should be equal to name.?", "name.?", arg.getName());
+            assertEquals("arg.getValue() should be equal to value_here", "value_ here", arg.getValue());
+            assertEquals("arg.getEncodedName() should be equal to name.%3F", "name.%3F", arg.getEncodedName());
+            assertEquals("arg.getValue() should be equal to value_+here", "value_+here", arg.getEncodedValue());
+            // Show that can bypass encoding:
+            arg.setAlwaysEncoded(false);
+            assertEquals("arg.getEncodedName() should be equal to name.?", "name.?", arg.getEncodedName());
```
Comment (vlsi) [inline test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java]: Again. The message does not clarify the intention of the test.
`should be equal to name.?` should be replaced with something like `setAlwaysEncoded=false => name.? should not be decoded` (I've no idea what setAlwaysEncoded does, and it is really hard to tell what the test is verifying)

File: test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java [hunk 5]
```diff
@@ -88,10 +96,120 @@ public void testEncoding() throws Exception {
...
+        @Test
+        public void testUnsupportedEncodingException() throws Exception {
+            // When invalid encoding(UTF-9) is passed error occurs
+            exception.expect(Error.class);
+            exception.expectMessage("java.io.UnsupportedEncodingException: UTF-9");
+            new HTTPArgument("name.?", "\00\01\07", true, "UTF-9");
+        }
+    
+        @Test
+        public void testWithFix() throws Exception {
```
Comment (vlsi) [inline test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java]: What does `testWithFix` mean?

**Review by vlsi — CHANGES_REQUESTED** [review]

File: test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java (hunk 5 above)
Comment (AbhaySinghNec) [reply test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java]: Renamed the "testWithFix" into meaningful name

File: test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java [hunk 6]
```diff
@@ -39,7 +47,7 @@ public void testCloning() throws Exception {
             assertEquals("name.?", clone.getName());
             assertEquals("value_ here", clone.getValue());
         }
-
+
```
[the remark this answers is not in the document]
Comment (AbhaySinghNec) [reply test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java]: Review comments incorporated

File: src/protocol/http/org/apache/jmeter/protocol/http/util/HTTPArgument.java [hunk 7]
```diff
@@ -146,19 +147,21 @@ public HTTPArgument(String name, String value, boolean alreadyEncoded, String co
...
                 }
-                value = URLDecoder.decode(value, contentEncoding);
+                value = urlCodec.decode(value, contentEncoding);
             } catch (UnsupportedEncodingException e) {
-                log.error("{} encoding not supported!", contentEncoding);
+                log.error(contentEncoding + " encoding not supported!");
                 throw new Error(e.toString(), e);
+            } catch (DecoderException e) {
+                log.error(contentEncoding + " exception in decoder!");
+                throw new IllegalArgumentException(e.toString(), e);
```
[the remark this answers is not in the document]
Comment (AbhaySinghNec) [reply src/protocol/http/org/apache/jmeter/protocol/http/util/HTTPArgument.java]: Review comments incorporated

File: src/protocol/http/org/apache/jmeter/protocol/http/util/HTTPArgument.java [hunk 8]
```diff
@@ -146,19 +147,21 @@ public HTTPArgument(String name, String value, boolean alreadyEncoded, String co
             try {
                 // We assume the name is always encoded according to spec
                 if(log.isDebugEnabled()) {
-                    log.debug("Decoding name, calling URLDecoder.decode with '{}' and contentEncoding: '{}'", name,
-                            EncoderCache.URL_ARGUMENT_ENCODING);
+                    log.debug("Decoding name, calling URLDecoder.decode with '"+name+"' and contentEncoding:"+EncoderCache.URL_ARGUMENT_ENCODING);
```
[the remark this answers is not in the document]
Comment (sanjeevnec) [reply src/protocol/http/org/apache/jmeter/protocol/http/util/HTTPArgument.java]: The unnecessary change has been removed.

File: src/protocol/http/org/apache/jmeter/protocol/http/util/HTTPArgument.java [hunk 9]
```diff
@@ -146,19 +147,21 @@ public HTTPArgument(String name, String value, boolean alreadyEncoded, String co
...
                 if(log.isDebugEnabled()) {
-                    log.debug("Decoding value, calling URLDecoder.decode with '{}' and contentEncoding: '{}'", value,
-                            contentEncoding);
+                    log.debug("Decoding value, calling URLDecoder.decode with '"+value+"' and contentEncoding:"+contentEncoding);
                 }
-                value = URLDecoder.decode(value, contentEncoding);
+                value = urlCodec.decode(value, contentEncoding);
             } catch (UnsupportedEncodingException e) {
-                log.error("{} encoding not supported!", contentEncoding);
+                log.error(contentEncoding + " encoding not supported!");
```
[the remark this answers is not in the document]
Comment (sanjeevnec) [reply src/protocol/http/org/apache/jmeter/protocol/http/util/HTTPArgument.java]: The unnecessary change has been removed.

File: src/protocol/http/org/apache/jmeter/protocol/http/util/HTTPArgument.java [hunk 10]
```diff
@@ -146,19 +147,21 @@ public HTTPArgument(String name, String value, boolean alreadyEncoded, String co
...
+                    log.debug("Decoding name, calling URLDecoder.decode with '"+name+"' and contentEncoding:"+EncoderCache.URL_ARGUMENT_ENCODING);
                 }
-                name = URLDecoder.decode(name, EncoderCache.URL_ARGUMENT_ENCODING);
+                URLCodec urlCodec = new URLCodec(contentEncoding);
+                name = urlCodec.decode(name, EncoderCache.URL_ARGUMENT_ENCODING);
                 // The value is encoded in the specified encoding
                 if(log.isDebugEnabled()) {
-                    log.debug("Decoding value, calling URLDecoder.decode with '{}' and contentEncoding: '{}'", value,
-                            contentEncoding);
+                    log.debug("Decoding value, calling URLDecoder.decode with '"+value+"' and contentEncoding:"+contentEncoding);
```
[the remark this answers is not in the document]
Comment (sanjeevnec) [reply src/protocol/http/org/apache/jmeter/protocol/http/util/HTTPArgument.java]: The unnecessary change has been removed.

File: test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java [hunk 11]
```diff
@@ -45,7 +45,7 @@ public void testCloning() throws Exception {
             assertEquals("name.%3F", clone.getEncodedName());
             assertEquals("value_+here", clone.getEncodedValue());
             assertEquals("name.?", clone.getName());
-            assertEquals("value_+here", clone.getValue());
+            assertEquals("value_ here", clone.getValue());
```
Comment (vlsi) [inline test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java]: Why is this change? Is it intentional?

File: test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java [hunk 12]
```diff
@@ -81,7 +81,7 @@ public void testEncoding() throws Exception {
             assertEquals("value_ here", arg.getValue());
             assertEquals("name.%3F", arg.getEncodedName());
             assertEquals("value_+here", arg.getEncodedValue());
-            arg.setAlwaysEncoded(false); // by default, name/value are encoded on fetch
+            arg.setAlwaysEncoded(false); // by default, name/value are encoded on fetch
```
Comment (vlsi) [inline test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java]: Is this change intentional?

File: test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java [hunk 13]
```diff
@@ -135,46 +135,44 @@ public void testUnsupportedEncodingException() throws Exception {
 
         @Test
         public void testWithShift_JIS_Encoding() throws Exception {
-        	
+
```
Comment (vlsi) [inline test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java]: Is this change intentional?

File: test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java (hunk 11 above)
Comment (sanjeevnec) [reply test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java]: This change was done in order to pass the test case testCloning. It wasn't supposed to be changed, was changed by mistake by us.

File: test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java (hunk 12 above)
Comment (sanjeevnec) [reply test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java]: This change was done in order to not include any unnecessary changes [trailing whitespace removal]. If you want I can make a different commit and remove this trailing whitespace. Please suggest.

File: src/protocol/http/org/apache/jmeter/protocol/http/util/HTTPArgument.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (sanjeevnec) [reply src/protocol/http/org/apache/jmeter/protocol/http/util/HTTPArgument.java]: The exception is only logged and not raised now.

File: test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java [hunk 14]
```diff
@@ -67,17 +75,17 @@ public void testEncoding() throws Exception {
             arg.setAlwaysEncoded(false);
             assertEquals("name.?", arg.getEncodedName());
             assertEquals("value_ here", arg.getEncodedValue());
-            
+
```
[the remark this answers is not in the document]
Comment (sanjeevnec) [reply test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java]: Apologies for that, will make sure doesn't happen next time.

File: test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java (hunk 3 above)
Comment (sanjeevnec) [reply test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java]: Assert messages have been updated as per your suggestion.

File: src/protocol/http/org/apache/jmeter/protocol/http/util/HTTPArgument.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (sanjeevnec) [reply src/protocol/http/org/apache/jmeter/protocol/http/util/HTTPArgument.java]: Turns out, if we do not raise an exception it breaks one of the test cases, hence raising it now instead of logging it.

File: test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java (hunk 13 above)
Comment (sanjeevnec) [reply test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java]: Created a new commit for these formatting related changes. Apologies for the unrelated change.

File: test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java (hunk 12 above)
Comment (sanjeevnec) [reply test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java]: Created a new commit to remove this trailing whitespace.

File: test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java (hunk 4 above)
Comment (sanjeevnec) [reply test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java]: setAlwaysEncoded(param) method changes the property AlwaysEncoded of the HTTPArgument. It is then checked by the getEncodedName() and getEncodedValue() methods and if isAlwaysEncoded() returns true then Name/Value is encoded else it isn't encoded. Attaching the implementation of getEncodedName()

`    public String getEncodedName() {
        if (isAlwaysEncoded()) {
            return cache.getEncoded(getName());
        } else {
            return getName();
        }`

hence updated the message `setAlwaysEncoded=false => arg.getEncodedName() should not be encoded`

File: test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java [hunk 15]
```diff
@@ -19,13 +19,21 @@
...
 
 import org.apache.jmeter.config.Arguments;
 import org.apache.jmeter.testelement.property.CollectionProperty;
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.rules.ExpectedException;
 
 public class TestHTTPArgument {
 
+        @Rule
```
Comment (ham1) [inline test/src/org/apache/jmeter/protocol/http/util/TestHTTPArgument.java]: Please use 4 spaces for indents

**Comment by pmouawad** [comment]
Thanks for this PR.
Still it is not mergeable as is and introduces too much risks of breaking existing tests, as it touches critical parts without covering it with Non Regression Testing.
I am closing it for now.
